# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/animation/KeyframeEffect.cpp
# Commit: 6f62586a139b
# Full Hash: 6f62586a139b1fcc4817dddd924400958d89a6bb
# Author: Zach Hoffman <zach@zrhoffman.net>
# Date: 2023-12-18 17:17:57
# Regressor Bug: 1846516
# File Overlap Count: 2
# Description:
#   Bug 1846516 - [css-properties-values-api] Animate custom properties in a discrete way. r=firefox-style-system-reviewers,emilio
#   
#   This patch also updates the bug ID for a FIXME leftover from bug 1840478
#   to bug 1869476, since the same FIXME is added in D190758.
#   
# ==============================================================================

diff -r 729abedf9bda -r 6f62586a139b dom/animation/KeyframeEffect.cpp
--- a/dom/animation/KeyframeEffect.cpp	Mon Dec 18 09:24:08 2023 +0000
+++ b/dom/animation/KeyframeEffect.cpp	Mon Dec 18 09:24:08 2023 +0000
@@ -27,6 +27,7 @@
 #include "mozilla/StaticPrefs_dom.h"
 #include "mozilla/StaticPrefs_gfx.h"
 #include "mozilla/StaticPrefs_layers.h"
+#include "nsCSSPropertyID.h"
 #include "nsComputedDOMStyle.h"  // nsComputedDOMStyle::GetComputedStyle
 #include "nsContentUtils.h"
 #include "nsCSSPropertyIDSet.h"
@@ -328,7 +329,8 @@
 bool KeyframeEffect::HasEffectiveAnimationOfPropertySet(
     const nsCSSPropertyIDSet& aPropertySet, const EffectSet& aEffectSet) const {
   for (const AnimationProperty& property : mProperties) {
-    if (aPropertySet.HasProperty(property.mProperty.mID) &&
+    if (!property.mProperty.IsCustom() &&
+        aPropertySet.HasProperty(property.mProperty.mID) &&
         IsEffectiveProperty(aEffectSet, property.mProperty.mID)) {
       return true;
     }
@@ -356,6 +358,9 @@
   AnimationPerformanceWarning::Type dummyWarning;
 
   for (const AnimationProperty& property : mProperties) {
+    if (property.mProperty.IsCustom()) {
+      continue;
+    }
     if (!compositorAnimatables.HasProperty(property.mProperty.mID)) {
       continue;
     }
@@ -467,6 +472,7 @@
   mCumulativeChanges = {};
   for (AnimationProperty& property : mProperties) {
     property.mIsRunningOnCompositor =
+        !property.mProperty.IsCustom() &&
         runningOnCompositorProperties.HasProperty(property.mProperty.mID);
     CalculateCumulativeChangesForProperty(property);
   }
@@ -590,7 +596,7 @@
       Servo_ComputedValues_ExtractAnimationValue(aBaseComputedStyle,
                                                  &aProperty.mProperty)
           .Consume();
-  mBaseValues.InsertOrUpdate(aProperty.mProperty.mID, std::move(baseValue));
+  mBaseValues.InsertOrUpdate(aProperty.mProperty, std::move(baseValue));
 }
 
 void KeyframeEffect::WillComposeStyle() {
@@ -628,7 +634,8 @@
     MOZ_ASSERT(prop.mSegments[prop.mSegments.Length() - 1].mToKey == 1.0,
                "incorrect last to key");
 
-    if (aPropertiesToSkip.HasProperty(prop.mProperty.mID)) {
+    if (!prop.mProperty.IsCustom() &&
+        aPropertiesToSkip.HasProperty(prop.mProperty.mID)) {
       continue;
     }
 
@@ -687,10 +694,13 @@
 
 void KeyframeEffect::SetIsRunningOnCompositor(nsCSSPropertyID aProperty,
                                               bool aIsRunning) {
+  MOZ_ASSERT(aProperty != eCSSPropertyExtra_variable,
+             "Can't animate variables on compositor");
   MOZ_ASSERT(
       nsCSSProps::PropHasFlags(aProperty, CSSPropFlags::CanAnimateOnCompositor),
-      "Property being animated on compositor is a recognized "
+      "Property being animated on compositor is not a recognized "
       "compositor-animatable property");
+
   for (AnimationProperty& property : mProperties) {
     if (property.mProperty.mID == aProperty) {
       property.mIsRunningOnCompositor = aIsRunning;
@@ -710,7 +720,8 @@
 void KeyframeEffect::SetIsRunningOnCompositor(
     const nsCSSPropertyIDSet& aPropertySet, bool aIsRunning) {
   for (AnimationProperty& property : mProperties) {
-    if (aPropertySet.HasProperty(property.mProperty.mID)) {
+    if (!property.mProperty.IsCustom() &&
+        aPropertySet.HasProperty(property.mProperty.mID)) {
       MOZ_ASSERT(nsCSSProps::PropHasFlags(property.mProperty.mID,
                                           CSSPropFlags::CanAnimateOnCompositor),
                  "Property being animated on compositor is a recognized "
@@ -1280,35 +1291,15 @@
       return;
     }
 
-    RefPtr<StyleLockedDeclarationBlock> customProperties;
-    // A workaround for CSS Animations in servo backend, custom properties in
-    // keyframe are stored in a servo's declaration block. Find the declaration
-    // block to resolve CSS variables in the keyframe.
-    // This workaround will be solved by bug 1391537.
-    if (isCSSAnimation) {
-      for (const PropertyValuePair& propertyValue : keyframe.mPropertyValues) {
-        if (propertyValue.mProperty.mID ==
-            nsCSSPropertyID::eCSSPropertyExtra_variable) {
-          customProperties = propertyValue.mServoDeclarationBlock;
-          break;
-        }
-      }
-    }
-
     JS::Rooted<JSObject*> keyframeObject(aCx, &keyframeJSValue.toObject());
     for (const PropertyValuePair& propertyValue : keyframe.mPropertyValues) {
       nsAutoCString stringValue;
-      // Don't serialize the custom properties for this keyframe.
-      if (propertyValue.mProperty.mID ==
-          nsCSSPropertyID::eCSSPropertyExtra_variable) {
-        continue;
-      }
       if (propertyValue.mServoDeclarationBlock) {
         Servo_DeclarationBlock_SerializeOneValue(
             propertyValue.mServoDeclarationBlock, &propertyValue.mProperty,
-            &stringValue, computedStyle, customProperties, rawData);
+            &stringValue, computedStyle, nullptr, rawData);
       } else {
-        if (auto* value = mBaseValues.GetWeak(propertyValue.mProperty.mID)) {
+        if (auto* value = mBaseValues.GetWeak(propertyValue.mProperty)) {
           Servo_AnimationValue_Serialize(value, &propertyValue.mProperty,
                                          rawData, &stringValue);
         }